Editable bike check in table

allenmichael099 opened the pull request:
Hi @asiripanich, I'm working on an editable bike check in table, but when I try to place the dtedit instructions in a module, I have trouble getting the data to display. I based my mod_DTedit off of mod_DT. I think my issue has something to do with the namespaces. This PR currently has the call to dtedit and the necessary callback function definitions placed in app_server.

In the call to dtedit, the name parameter defines the name of the object available to the uiOutput. 
Within dtedit, renderUI is called like this 'output[[name]] <- shiny::renderUI', and dtedit uses DT::renderDataTable({. 

More explanation is in the readme:
https://github.com/jbryer/DTedit

And I have been looking off of this example:
https://github.com/jbryer/DTedit/blob/master/inst/shiny_demo/app.R

allenmichael099 committed:
Bike check In Table

- Added function query_checkinout to utils_query_databasee
- Added the checkinout field to data_r in mod_load_data
- called mod_DT_server using data = data_r$checkinout
-  added a tabPanel for "Bike Check In" to app_ui
- moved the assert_flag for anon_locations within run_app

allenmichael099 committed:
Dynamically adding tables

attempted to use renderUI and uiOutput

allenmichael099 committed:
Programmatic Table Addition

app_server:
- looped over each table
- used appendTab within app_server to add each new table, then called mod_DT_server for it
- moved normalize uuid and setcolorder from query_supplementary to the line after query supplementary was called within mod_load_data

allenmichael099 committed:
Moved config to inst

allenmichael099 committed:
Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table

allenmichael099 committed:
Revert "Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table"

This reverts commit 3f9cbb1062dd052e233e40e0a871a9cacca8f690, reversing
changes made to 36636e522f3e165f2e200c19b55ee3adf1250692.

allenmichael099 committed:
Revert "Revert "Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table""

This reverts commit 15c7fa52e2e0547767454edbe86b2251e8c2f48e.

allenmichael099 committed:
Small Fixes

- Removed Suggestion and Tier Sys from the default config and commented them out of app_config.yml

-  mod_load_data: Removed extra load participants and the extra generate_trajectories comment

- utils_query_database: Changed confirmed_trip to cleaned_trip within query_cleaned_trips

allenmichael099 committed:
cleaned_trip -> confirmed_trip

- within query_cleaned_trips, changed cleaned_trip to confirmed_trip
- commented out polar bear in the default yaml config

allenmichael099 committed:
Delete app_config.yml

allenmichael099 committed:
Editable Bike Check In

- Commented out all displays except tables
-  the data tabke for bike check in gets rendered by dtedit
- attempted to make a module for dtedit similar to mod_DT

allenmichael099 commented:
I also commented out the other displays to focus on the tables.

allenmichael099 commented:
What do you think I should do?

asiripanich commented:
Hi @allenmichael099, I think it is a good idea not to push your test code with your PR (i.e., `DTedit_checkinout_app.R`). It can be quite confusing.

I tried to run `DTedit_checkinout_app.R` but got an error saying the input is empty.

Does `get_Data()` return a non-emptied table for you?

asiripanich commented:
Try this and let me know if it works for you or not.

```r
library(shiny)
library(DTedit)
#> Loading required package: DT
#> 
#> Attaching package: 'DT'
#> The following objects are masked from 'package:shiny':
#> 
#>     dataTableOutput, renderDataTable
ui <- function(id) {
  fluidPage(uiOutput("DTedit_table"))
}

server <- function(input, output, session) {
  bike_data <- cars

  insert_callback <- function(data, row) {
    data_template <- data.frame(speed = NA, dist = NA)
    data <- rbind(data, data_template)
    return(data)
  }

  update_callback <- function(data, olddata, row) {
    data
  }

  delete_callback <- function(data, row) {
    return(data[-row, ])
  }

  message(class(bike_data))
  output$DTedit_table <- renderUI({
    DTedit::dtedit(input, output,
      name = "DTedit_table",
      thedata = bike_data,
      edit.cols = c("speed", "dist"),
      input.types = c(speed = "textInput"),
      input.choices = list(dist = 1:100),
      view.cols = c("speed", "dist"),
      callback.update = update_callback,
      callback.insert = insert_callback,
      callback.delete = delete_callback
    )
  })
}

shinyApp(ui, server)
#> PhantomJS not found. You can install it with webshot::install_phantomjs(). If it is installed, please make sure the phantomjs executable can be found via the PATH variable.
```

asiripanich commented on R/app_server.R:
```diff
@@ -125,15 +125,89 @@ app_server <- function(input, output, session) {
+    # Define the callback functions used by dtedit
+    #' Insert a row. "Create"
+    insert_callback <- function(data, row) {
+      mydata <- rbind(data, mydata)
```
`mydata` is not defined.

asiripanich commented on R/app_server.R:
```diff
@@ -125,15 +125,89 @@ app_server <- function(input, output, session) {
+    
+    #' Delete a row
+    delete_callback <- function(data, row) {
+      mydata[row,] <- NULL
```
Try `mydata[-row, ]`

asiripanich commented on R/app_server.R:
```diff
@@ -125,15 +125,89 @@ app_server <- function(input, output, session) {
+      #            id = paste0("DT_ui_",table_type),
+      #            data = data_r[[table_type]] )
+      
+      DTedit::dtedit(input, output,
```
The `Checkinout` table is empty. So I don't think this code would work.

asiripanich reviewed (commented):
Initial review..

allenmichael099 commented:
Sorry about that, I forgot I had DTedit_checkinout_app.R in there. Also I didn't think about the fact that the Checkinout data is not in the test data - I had to add it. The script you shared with cars data works on its own for me, but leads to this warning:
Warning: Error in as.vector: cannot coerce type 'environment' to vector of type 'character'. I think it has to do with the server calling renderUI for dtedit, which already calls renderUI.

Using your cars example, I made an example of the issue I'm having. When I place DTedit inside server and uiOutput inside ui, the table displays. When I move DTedit to mod_DTedit_server and uiOutput to mod_DTedit_ui, the editing buttons appear but not the table.

DTedit inside main server:
![image](https://user-images.githubusercontent.com/56986329/121709402-c0e36d00-ca95-11eb-9011-8ec5a1d3a260.png)

DTedit inside module:
![image](https://user-images.githubusercontent.com/56986329/121709152-7c57d180-ca95-11eb-9615-d9d8dac694ef.png)

allenmichael099 commented:
Here is my example that leads to the second picture above.
```

mod_DTedit_ui <- function(id) {
  ns <- NS(id)
  uiOutput(outputId = ns("DTedit_table"))
}

mod_DTedit_server <- function(input, output, session, data) {
  ns <- session$ns 
  req(data)
  
  insert_callback <- function(data, row) {
    data_template <- data.frame(speed = NA, dist = NA)
    data <- rbind(data, data_template)
    return(data)
  }
  
  update_callback <- function(data, olddata, row) {
    data
  }
  
  delete_callback <- function(data, row) {
    return(data[-row, ])
  }
  
  # To check that the data is there and doesn't have 0 rows.
  message(class(data))
  dput(dim(data))
  DTedit::dtedit(input, output,
                 name = "DTedit_table",
                 thedata = data,
                 edit.cols = c("speed", "dist"),
                 input.types = c(speed = "textInput"),
                 input.choices = list(dist = 1:100),
                 view.cols = c("speed", "dist"),
                 callback.update = update_callback,
                 callback.insert = insert_callback,
                 callback.delete = delete_callback
    )
}

ui <- function(id) {
  fluidPage(mod_DTedit_ui("DTedit_table"))
}

server <- function(input, output, session) {
  
  bike_data <- cars
  
  callModule(module = mod_DTedit_server,
             id = "DTedit_table",
             data = bike_data)
}

shinyApp(ui, server)
```

allenmichael099 committed:
Delete DTedit_checkinout_app.R

asiripanich commented:
> Here is my example that leads to the second picture above.
> 
> ```
> 
> mod_DTedit_ui <- function(id) {
>   ns <- NS(id)
>   uiOutput(outputId = ns("DTedit_table"))
> }
> 
> mod_DTedit_server <- function(input, output, session, data) {
>   ns <- session$ns 
>   req(data)
>   
>   insert_callback <- function(data, row) {
>     data_template <- data.frame(speed = NA, dist = NA)
>     data <- rbind(data, data_template)
>     return(data)
>   }
>   
>   update_callback <- function(data, olddata, row) {
>     data
>   }
>   
>   delete_callback <- function(data, row) {
>     return(data[-row, ])
>   }
>   
>   # To check that the data is there and doesn't have 0 rows.
>   message(class(data))
>   dput(dim(data))
>   DTedit::dtedit(input, output,
>                  name = "DTedit_table",
>                  thedata = data,
>                  edit.cols = c("speed", "dist"),
>                  input.types = c(speed = "textInput"),
>                  input.choices = list(dist = 1:100),
>                  view.cols = c("speed", "dist"),
>                  callback.update = update_callback,
>                  callback.insert = insert_callback,
>                  callback.delete = delete_callback
>     )
> }
> 
> ui <- function(id) {
>   fluidPage(mod_DTedit_ui("DTedit_table"))
> }
> 
> server <- function(input, output, session) {
>   
>   bike_data <- cars
>   
>   callModule(module = mod_DTedit_server,
>              id = "DTedit_table",
>              data = bike_data)
> }
> 
> shinyApp(ui, server)
> ```

Hi @allenmichael099, can you please check the UI code that it is correct. See one of the existing UI mods, how we should write it.

allenmichael099 commented:
I think mod_DTedit_ui above follows the same structure as mod_DT_ui from emdash. The only major difference is DT::dataTableOutput vs uiOutput. The dtedit function calls  DT::dataTableOutput within renderUI. The spinner doesn't seem to impact whether the table appears. 
```
mod_DT_ui <- function(id) {
  ns <- NS(id)
  tagList(
    shinycssloaders::withSpinner(DT::dataTableOutput(ns("DTtable")))
  )
}
```

allenmichael099 committed:
Database editing functions

- Made update, insert, and delete methods to interact with the database
- UI not quite behaving right, need to look at the example more

allenmichael099 committed:
CUD Functions

- moved update, insert, and delete callback functions to utils_update_insert_delete.R
- set bike check in as editable with the delete option only
- added the csv, pdf, etc buttons back into bike check in

allenmichael099 commented:
"Bike Check In" now allows for deletion. I left the other parts of the dashboard commented for easier testing, but if it's easier to merge I can uncomment them. The commented out portions are in app_ui and app_server.

allenmichael099 committed:
Delete Allowed and Datetime display

- moved callback functions back to the server, since they require cons to exist before they get created
- added a column to the Bike Check In display: Checkout_time. This uses the system's timezone.
- uncommented the other portions of the dashboard

allenmichael099 commented:
@shankari Delete should work on this branch, and the checkout time is now displayed with the dashboard viewer's computer timezone. Checkout_time is a character string.
![image](https://user-images.githubusercontent.com/56986329/123891142-b1797600-d915-11eb-9834-99bac69e8fbe.png)

shankari commented on R/app_server.R:
```diff
@@ -125,15 +125,114 @@ app_server <- function(input, output, session) {
+      table_title <- t[[table_type]]$tab_name
+      
+      # If the table is Bike Check In, use dtedit
+      if (table_type == 'Checkinout') {
+        
+        # Define the callback functions used by dtedit
+        #' Insert a row. "Create"
+        #' @param data the data including your inserted row
+        #' @param row the row where you made the change
+        insert_callback <- function(data, row) {
+          db_insert(cons,"Checkinout",data[row,])
+          return(data)
+        }
+        
+        #' Update a row
+        #' @param data the data including your updated row
+        update_callback <- function(data, olddata, row) {
+          db_update(cons,"Checkinout",data[row,])
+          data
+        }
+        
+        #' Delete a row
+        delete_callback <- function(data, row) {
+          db_delete(cons,"Checkinout",data[row,])
+          return(data[-row, ])
+        }
+        
```
I will deploy this for now, but I was hoping that this would also be configurable. So as part of the config file, we would have a flag which said that the supplementary table was "editable". If it was, we would wrap it in an editable `DT_edit`. Note that we already have the table name in the config, can't we just use the table name here instead of hardcoding `Checkinout`?

asiripanich replied:
This should be made into a module, but for some reason we couldn't make it work with DTedit (if I remember this correctly).

shankari replied:
@asiripanich there is a more recently updated fork of DTEdit that claims that it is modularized (https://github.com/DavidPatShuiFong/DTedit). Maybe @allenmichael099 can look into that?
The same user submitted a PR to the main DTEdit, but it hasn't been merged yet (https://github.com/jbryer/DTedit/pull/5)

shankari commented on R/app_server.R:
```diff
@@ -125,15 +125,114 @@ app_server <- function(input, output, session) {
+                       edit.label.cols = c('status'),
+                       input.types = c(status = "selectInput"),
+                       input.choices = list(status = c('TRUE','FALSE')),
+                       view.cols = c('user_id','status', 'bikeLabel','ts','Checkout_time'),
```
again, can the cols be determined from the input data instead of being hardcoded?

asiripanich replied:
I agree, I think Michael was just in a rush to finish this. Now that he is back he shall revise this. :)

shankari commented:
@allenmichael099 there is a merge conflict in this PR

allenmichael099 committed:
Merge branch 'master' of https://github.com/asiripanich/emdash into Editable_Bike_Check_In_Table#

shankari commented:
When I try to deploy this, I get the error

```
Warning: Error in loadNamespace: there is no package called ‘DTedit’
  86: <Anonymous>
```

Does the dockerfile need to to be updated to include the new dependency?

allenmichael099 committed:
import DTedit

Added import DTedit to app_server and ran devtools::document()

allenmichael099 committed:
Add DTedit to dockerfile

shankari commented:
Now I'm getting the error

```
Rscript -e 'remotes::install_version("DTedit",upgrade="never",version = "1.0.0")'
Error in package_find_repo(package, repos) :
  couldn't find package 'DTedit'
Calls: <Anonymous> -> download_version_url -> package_find_repo
Execution halted
```

allenmichael099 committed:
Use install_github for DTedit

allenmichael099 committed:
Update DTedit view.cols

asiripanich commented:
I will fix this PR and merge.

DTEdit is not on CRAN. We should clone it and let emdash install the clone.

asiripanich committed:
Document

asiripanich committed:
Add DTedit to remotes

asiripanich committed:
Replace glue with sprintf

There are too many dependencies right now.

asiripanich committed:
Remove xml2

asiripanich committed:
Add DTedit to Imports

asiripanich committed:
Refactor

asiripanich commented:
Can you share the test database that you are using for this bike checkin feature? I'm happy to have a look to make sure it works.

shankari commented:
@allenmichael099 I deployed this, and it looks like the formatted date is based on the server timestamp, not the browser timestamp. The R server on production is running on AWS, so the formatted timestamp is in UTC although my browser is set to PDT.

<img width="1474" alt="Screen Shot 2021-07-02 at 11 06 06 AM" src="https://user-images.githubusercontent.com/2423263/124313289-86657100-db25-11eb-993b-ad8083f57647.png">

You probably didn't notice this since your R/Shiny server is also running on your laptop. Not sure there is a great way to test this during development other than using your personal laptop (presumably set to EDT) to access the dashboard that is running on your work laptop.

allenmichael099 committed:
Made 'editable' configurable

- added fields to Checkinout in the config
- which buttons appear is now configurable
- Checkinout will be static if editable is not a field

allenmichael099 committed:
Update utils_update_insert_delete.R

allenmichael099 committed:
Display Client Time

- used Javascript in app_UI to get the client's timezone offset
- added time_zone_offset argument to mod_load_data_server and query_supplementary
- to convert the timestamp to the client's time, subtract the offset in seconds from the timestamp
- the offset is positive if behind UTC, negative if ahead of UTC

allenmichael099 committed:
Update app_ui.R

allenmichael099 committed:
moved DTedit to mod_DTedit

allenmichael099 committed:
Revert "Update app_ui.R"

This reverts commit 6c3aa871efd87636725fa0bd9fe070e443f41d50.

allenmichael099 committed:
Revert "Display Client Time"

This reverts commit 16c77f71c432dd0c1869afe199a837fff9714125.

shankari commented:
your approach to fomatting in the browser timezone reads the time offset so that it can convert the timestamp on the server (R) side. I don't think this is a great option because:
- it will not display the timezone (e.g. PDT), so the user has no way to confirm what timezone the date was formatted in
- it will not work at the boundaries - e.g. if a bike was checked out at 4pm PST, after daylight savings starts, it will show as 3pm (PDT), which is fairly confusing if we want to ask the user about a particular bike.

An alternative is to use a session-based mechanism where the R sends the timestamp to javascript and javascript returns the formatted value. I double-checked that the built-in timezone does handle standard and daylight savings correctly.

```
new Date()
Date Tue Jul 06 2021 13:40:40 GMT-0700 (Pacific Daylight Time)

Date.now() - (6 * 30.5 * 24 * 60 * 60 * 1000)
1609792926120

new Date(1609792926120)
Date Mon Jan 04 2021 12:42:06 GMT-0800 (Pacific Standard Time)
```

An alternate approach is to use javascript with Datatables directly.
Apparently, there is a "callback" that we can create (written completely in javascript) that can allow us to modify an R DTTable (e.g. https://rstudio.github.io/DT/002-rowdetails.html). DTTables uses the javascript DataTables library, which in turn has the concept of a renderer which will display the underlying data in different formats.

https://datatables.net/manual/data/renderers

So theoretically, the R datatable can have "ts" and "fmt_time" columns and we can use the `callback` to add a renderer that wraps the values in `fmt_time` with Date.

```
{
    data: 'fmt_time',
    render: function ( data, type, row ) {
        return new Date(data * 100);
    }
}
```

shankari commented:
Aha! There is a way to invoke the rendering directly from the DT initializer!!!
https://rstudio.github.io/DT/options.html (section "Column rendering")

We should be able to define a function on the `fmt_time` column using the appropriate target (https://datatables.net/reference/option/columnDefs)

allenmichael099 commented:
@shankari Here is my attempt to render fmt_time as a date. It seems like the Javascript got ignored - the fmt_time showed up the same as ts.
```
library(DT)
# The function 'JS' is from the htmlwidgets library, which DT loads.

ui <- function(id) {
  fluidPage(tagList(
    shinycssloaders::withSpinner(DT::dataTableOutput("DTtable"))
  ))
}

server <- function(input, output, session) {
  bike_data <- cars[1:10,]
  
  # Make time stamp and fmt_time columns
  bike_data[['ts']] <- rep(as.numeric(Sys.time()),10)
  bike_data[['fmt_time']] <- bike_data[['ts']]
  fmt_time_index <- which(names(bike_data) == 'fmt_time') - 1  # subtract 1 because the index starts from 0

  output$DTtable <- DT::renderDataTable({
    DT::datatable(
      bike_data,
      autoHideNavigation = TRUE,
      options = list(
        scrollX = TRUE,
        pageLength = 3,
        dom = "Bfrtip",
        buttons = c("copy", "csv", "excel", "pdf", "print", "colvis"),
        
        ####### My attempt to render fmt_time as a date.
        columnDef = list(list(
          targets = fmt_time_index,   
          render = JS(
            "function(data, type, row) {",
            "return new Date(data*1000);",
            "}")
        ))
      ),
      filter = list(position = "top", clear = FALSE),
      extensions = "Buttons"
    )
  })
}

shinyApp(ui, server)
```

shankari commented:
@allenmichael099 this is a bit of an obvious fix, but shouldn't this be `columnDefs` (with a trailing **s** instead)

```
        ####### My attempt to render fmt_time as a date.
        columnDef = list(list(
```

If that is not the issue, maybe you can try some of the sample renderers (https://datatables.net/manual/data/renderers) with a hardcoded target and see if you can get any of the columns to work.

allenmichael099 committed:
Render fmt_time

- added columnDefs argument to datatable options in app_server
- added DT_options argument to mod_DT
- moved ts column manipulation from query_supplementary to app_server

allenmichael099 committed:
Merge branch 'master' of https://github.com/asiripanich/emdash into Editable_Bike_Check_In_Table

allenmichael099 committed:
Delete trips_with_empty_user_input.Rdata

allenmichael099 committed:
Import htmlwidgets

- I'm not sure how to handle this. DT uses htmlwidgets, but when I don't specify htmlwidgets::, the function 'JS' isn't recognized

allenmichael099 commented:
@asiripanich I'm not sure I imported htmlwidgets correctly. DT loads htmlwidgets, but without either specifying `htmlwidgets::JS` or making a `library()` call to either htmlwidgets or DT, the function 'JS' isn't recognized. I use 'JS' within app_server to set a columnDefs option for rendering any given supplementary table's fmt_time as a date.

allenmichael099 committed:
Added htmlwidgets to Imports in the description

Ran usethis::use_package('htmlwidgets')

shankari commented:
@asiripanich what is holding up this merge?

asiripanich commented:
It was me, I guess. :( 

Just to confirm, you only want to allow the user to remove rows? why not "add" and "modify" too?

![image](https://user-images.githubusercontent.com/17020181/126244592-7cbb8a87-7ad5-4aeb-a51c-ee0ba3ba0821.png)

Also, are we not using the DTEdit fork that can be modularised?


Happy to merge as soon as those questions are answered. :)

asiripanich committed:
Remove unnecessary `return()`

allenmichael099 committed:
Fixed table duplication when reloading

- moved append tab and dtedit to the observe event for reloading participants
- made the variable 'startup_load_supplementary' to signal when to append a tab

allenmichael099 committed:
Remove extra messages and browser() statement

allenmichael099 committed:
Prep for merge with Editable Bike Check In

- move rendering to inside mod_DTedit
- add DT_options arg. to mod_DT

allenmichael099 committed:
Merge branch 'modular_DTedit' into Editable_Bike_Check_In_Table

allenmichael099 committed:
Switched version of DTedit to install

- now using the modular version

allenmichael099 commented:
If for some reason there is an issue such as 'unrecognized function dteditmodui', I recommend using this commit:
`Remove extra messages and browser() statement`, which should work fine.

So far the modularized version is working for me.

allenmichael099 commented:
@asiripanich I left out insert because there is a potential problem of an entry being created that doesn't have user id, or doesn't have a bikeLabel. Update is left out because if all status' are TRUE, the option 'FALSE' will not be presented for update.

asiripanich committed:
Fix document

asiripanich commented:
> So far the modularized version is working for me.

This is probably because the version of DTedit that emdash currently depends on is my fork.

asiripanich commented:
> @asiripanich I left out insert because there is a potential problem of an entry being created that doesn't have user id, or doesn't have a bikeLabel. Update is left out because if all status' are TRUE, the option 'FALSE' will not be presented for update.

Just to confirm again, @allenmichael099 and @shankari are happy with just the "delete" option?

asiripanich committed:
Merge branch 'Editable_Bike_Check_In_Table' of github.com:allenmichael099/emdash into Editable_Bike_Check_In_Table

asiripanich committed:
Use a modulisable fork of DTedit

asiripanich commented:
Also, is it ok if I add the `Checkinout.bson` you emailed me to the test database?

allenmichael099 commented:
Yep! The delete option and the  Checkinout.bson I sent are fine.

asiripanich committed:
Styler

asiripanich commented:
Will merge in a few hours. :)

asiripanich committed:
- Add a test checkinout table to the test database

shankari commented:
@asiripanich I am fine with only delete for this feature

allenmichael099 committed:
Document the arguments for db_update/insert/delete

allenmichael099 committed:
Document DTedit server parameters

- and converted callback function roxygen comments to regular comments

allenmichael099 committed:
Ran devtools::document()

asiripanich committed:
Fix document

asiripanich committed:
Refactor

asiripanich committed:
Refactor

- `query_supplementary()` is very misleading and it is just a simple query from a collection. I think it is better not to define a new function for it.

asiripanich committed:
Cleaning

asiripanich committed:
Fix

asiripanich committed:
Update Dockerfile
